tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp

```diff
@@ -0,0 +1,297 @@
+
+  MAKE_GENERATOR(generator);
+  std::uniform_real_distribution<> dist(-1.0, 1.0);
+  std::uniform_real_distribution<> dist_positive(1.0, 2.0);
+  std::uniform_real_distribution<> dist_pert(-0.1, 0.1);
+  const auto nn_generator = make_not_null(&generator);
+  const auto nn_dist = make_not_null(&dist);
+  const auto nn_dist_pert = make_not_null(&dist_pert);
+
+  const auto box = db::create<
+      db::AddSimpleTags<StrahlkorperTags::items_tags<Frame::Inertial>>,
+      db::AddComputeTags<
+          StrahlkorperTags::compute_items_tags<Frame::Inertial>>>(strahlkorper);
+  const auto& unnormalized_normal =
+      db::get<StrahlkorperTags::NormalOneForm<Frame::Inertial>>(box);
+
+  const auto one =
+      make_with_value<Scalar<DataVector>>(unnormalized_normal, 1.0);
+  const auto minus_one =
+      make_with_value<Scalar<DataVector>>(unnormalized_normal, -1.0);
+  const auto five =
+      make_with_value<Scalar<DataVector>>(unnormalized_normal, 5.0);
+  const auto minus_five =
+      make_with_value<Scalar<DataVector>>(unnormalized_normal, -5.0);
+
+  // Choose spacetime_metric randomly, but make sure the result is
+  // still invertible. To do this, start with
+  // Minkowski, and then add a 10% random perturbation.
+  auto spacetime_metric_int = make_with_random_values<
+      tnsr::aa<DataVector, spatial_dim, Frame::Inertial>>(
+      nn_generator, nn_dist_pert, unnormalized_normal);
+  get<0, 0>(spacetime_metric_int) += get(minus_one);
+  get<1, 1>(spacetime_metric_int) += get(one);
+  get<2, 2>(spacetime_metric_int) += get(one);
+  get<3, 3>(spacetime_metric_int) += get(one);
+
+  auto spacetime_metric_ext = make_with_random_values<
+      tnsr::aa<DataVector, spatial_dim, Frame::Inertial>>(
+      nn_generator, nn_dist_pert, unnormalized_normal);
+  get<0, 0>(spacetime_metric_ext) += get(minus_one);
+  get<1, 1>(spacetime_metric_ext) += get(one);
+  get<2, 2>(spacetime_metric_ext) += get(one);
+  get<3, 3>(spacetime_metric_ext) += get(one);
+
+  // Set pi, phi to be random (phi, pi should not need to be consistent with
+  // spacetime_metric for the flux consistency tests to pass)
+  const auto phi_int = make_with_random_values<
+      tnsr::iaa<DataVector, spatial_dim, Frame::Inertial>>(
+      nn_generator, nn_dist_pert, unnormalized_normal);
+  const auto pi_int = make_with_random_values<
+      tnsr::aa<DataVector, spatial_dim, Frame::Inertial>>(
+      nn_generator, nn_dist_pert, unnormalized_normal);
+  const auto phi_ext = make_with_random_values<
+      tnsr::iaa<DataVector, spatial_dim, Frame::Inertial>>(
+      nn_generator, nn_dist_pert, unnormalized_normal);
+  const auto pi_ext = make_with_random_values<
+      tnsr::aa<DataVector, spatial_dim, Frame::Inertial>>(
+      nn_generator, nn_dist_pert, unnormalized_normal);
+
+  const auto& spatial_metric_int = gr::spatial_metric(spacetime_metric_int);
+  const auto& inverse_spatial_metric_int =
+      determinant_and_inverse(spatial_metric_int).second;
```

Comment:
c++ automatically extends the lifetime of the temporary object returned by determinant_and_inverse to the lifetime of the local const reference. If I understand this blog post correctly, the destructors are also called correctly when the const ref does go out of scope.

https://blog.galowicz.de/2016/03/23/const_reference_to_temporary_object/

So I think this is fine, but I would love to understand why not if I'm wrong!

Reply:
While it's technically not wrong in this case I view it as writing misleading code. The code as written does not correctly portray what you are intending, which clearly leads to confusion. Furthermore, there are cases (as pointed out in the article you linked) where very similar looking code leads to incorrect behavior. I believe if `.second` were a member function you'd have UB. I've generally been asking people to not do what you're doing here because of these subtleties.

Reply:
OK, I removed the & here and below where I call det and inv